### PR TITLE
[Utility] Add skill framework

### PR DESCRIPTION
**Pull request by XuGW-Kevin**
设计了技能展示界面。已经与之前的PR进行了rebase。这次PR对自己的坦克有较小改动，这仅仅是为了适配技能展示界面。

### CODE REVIEW

File: src/battle_game/core/bullets/rocket.cpp [hunk 1]
```diff
@@ -14,29 +16,88 @@ Rocket::Rocket(GameCore *core,
                glm::vec2 velocity)
     : Bullet(core, id, unit_id, player_id, position, rotation, damage_scale),
       velocity_(velocity) {
+  auto &units = game_core_->GetUnits();
+  auto player = game_core_->GetPlayer(player_id_);
+  auto &input_data = player->GetInputData();
+  auto WhereCursorIs = input_data.mouse_cursor_position;
```
Comment (LazyJazz) [inline src/battle_game/core/bullets/rocket.cpp:22]: variable names should be in lower_case format

File: src/battle_game/core/bullets/rocket.cpp [hunk 2]
```diff
@@ -14,29 +16,88 @@ Rocket::Rocket(GameCore *core,
...
+      diff = glm::vec2{0.25 * harmful_ * diff.x / length,
+                       0.25 * harmful_ * diff.y / length};
+    }
+  }
+  velocity_ = diff;
   position_ += velocity_ * kSecondPerTick;
+  float angle = 0;
+  if (diff.x < 0) {
+    if (diff.y < 0) {
+      angle = 3.141592653 / 2 + atan(abs(diff.y) / abs(diff.x));
```
Comment (LazyJazz) [inline src/battle_game/core/bullets/rocket.cpp:67]: maybe you can use glm::pi<float>() instead, or you can add a macro somewhere public

File: src/battle_game/core/unit.h [hunk 3]
```diff
@@ -57,10 +57,14 @@ class Unit : public Object {
 
   [[nodiscard]] virtual const char *UnitName() const;
   [[nodiscard]] virtual const char *Author() const;
+  std::vector<Skill> GetSkill() const {
```
Comment (LazyJazz) [inline src/battle_game/core/unit.h:60]:
```suggestion
  const std::vector<Skill>& GetSkill() const {
```

File: src/battle_game/core/units/inferno_tank.cpp [hunk 4]
```diff
@@ -22,34 +45,43 @@ void InfernoTank::Update() {
   Fire();
 }
 
+void InfernoTank::Hidden_CLICK() {
+  isHidden = 3 * kTickPerSecond;
+  hidden_count_down_ = 6 * kTickPerSecond;
+}
+void InfernoTank::Block_CLICK() {
```
Comment (LazyJazz) [inline src/battle_game/core/units/inferno_tank.cpp:52]: function names should be in CamelCase

**Comment by LazyJazz** [comment]
Could you please upload some screenshots?

**Review by LazyJazz — CHANGES_REQUESTED** [review]
Good job.

File: src/battle_game/app/app.cpp [hunk 5]
```diff
@@ -271,11 +273,90 @@ void App::UpdateImGui() {
       }
       auto unit = game_core_->GetUnit(player->GetPrimaryUnitId());
       if (unit) {
-        ImGui::Text("Health: %.1f / %.1f",
+        ImGui::Text("生命值: %.1f / %.1f",
```
[the remark this answers is not in the document]
Comment (XuGW-Kevin) [reply src/battle_game/app/app.cpp:276]: Thanks. Fixed

File: src/battle_game/core/bullets/rocket.cpp (hunk 1 above)
Comment (XuGW-Kevin) [reply src/battle_game/core/bullets/rocket.cpp:22]: Thanks. Fixed

File: src/battle_game/core/bullets/rocket.cpp (hunk 2 above)
Comment (XuGW-Kevin) [reply src/battle_game/core/bullets/rocket.cpp:67]: Thanks. Fixed

File: src/battle_game/core/unit.h (hunk 3 above)
Comment (XuGW-Kevin) [reply src/battle_game/core/unit.h:60]: Thanks. Fixed.

File: src/battle_game/core/units/inferno_tank.cpp (hunk 4 above)
Comment (XuGW-Kevin) [reply src/battle_game/core/units/inferno_tank.cpp:52]: Thanks. Fixed

**Comment by XuGW-Kevin** [comment]
example

File: src/battle_game/core/game_core.h [hunk 6]
```diff
@@ -32,6 +32,9 @@ class GameCore {
   void GeneratePrimaryUnitList();
   uint32_t AllocatePrimaryUnit(uint32_t player_id);
   [[nodiscard]] std::vector<const char *> GetSelectableUnitList() const;
+  [[nodiscard]] std::vector<bool> GetSelectableUnitListSkill() const {
```
Comment (LazyJazz) [inline src/battle_game/core/game_core.h:35]:
```suggestion
  [[nodiscard]] const std::vector<bool>& GetSelectableUnitListSkill() const {
```

File: src/battle_game/core/game_core.h [hunk 7]
```diff
@@ -203,6 +206,7 @@ class GameCore {
   std::vector<std::function<uint32_t(uint32_t)>>
       primary_unit_allocation_functions_;
   std::vector<std::string> selectable_unit_list_;
+  std::vector<bool> selectable_unit_list_skill_;
```
Comment (LazyJazz) [inline src/battle_game/core/game_core.h:209]: What is the purpose of this list? Please add some explanation in comments

File: src/battle_game/core/unit.h [hunk 8]
```diff
@@ -57,10 +57,14 @@ class Unit : public Object {
 
   [[nodiscard]] virtual const char *UnitName() const;
   [[nodiscard]] virtual const char *Author() const;
+  const std::vector<Skill> &GetSkill() const {
```
Comment (LazyJazz) [inline src/battle_game/core/unit.h:60]:
```suggestion
  const std::vector<Skill> &GetSkills() const {
```

File: src/battle_game/core/unit.h [hunk 9]
```diff
@@ -57,10 +57,14 @@ class Unit : public Object {
...
+    return skill_;
```
Comment (LazyJazz) [inline src/battle_game/core/unit.h:61]:
```suggestion
    return skills_;
```

File: src/battle_game/core/unit.h [hunk 10]
```diff
@@ -57,10 +57,14 @@ class Unit : public Object {
...
+  }
 
  protected:
   uint32_t player_id_{};
   float health_{1.0f};
+  std::vector<Skill> skill_;
```
Comment (LazyJazz) [inline src/battle_game/core/unit.h:67]:
```suggestion
  std::vector<Skill> skills_;
```

**Review by LazyJazz — CHANGES_REQUESTED** [review]
Still need some changes. Thanks.

**Comment by XuGW-Kevin** [comment]
OK! Wait a minute and all the changes will be done as soon as possible! Thanks!!!!

File: src/battle_game/core/game_core.h (hunk 7 above)
Comment (XuGW-Kevin) [reply src/battle_game/core/game_core.h:209]: In case someone do not want their skills to be displayed.

**Comment by XuGW-Kevin** [comment]
Fixed. Thanks a lot!

**Comment by XuGW-Kevin** [comment]
Thank you @LazyJazz . Now all done and you may feel free to check a little more.

**Review by LazyJazz — APPROVED** [review]
LGTM!

**Pull request merged**